### PR TITLE
reduce physics simulation cost of other-avatars

### PR DESCRIPTION
**Pull request by AndrewMeadows**
- ignore other-avatar angularVelocity in physics simulation
- ignore other-avatar collisions against static and kinematic objects
- more correct other-avatar diameter when slaving linear motion

### CODE REVIEW

File: interface/src/avatar/AvatarMotionState.cpp [hunk 1]
```diff
@@ -19,6 +19,7 @@
 AvatarMotionState::AvatarMotionState(AvatarSharedPointer avatar, const btCollisionShape* shape) : ObjectMotionState(shape), _avatar(avatar) {
     assert(_avatar);
     _type = MOTIONSTATE_TYPE_AVATAR;
+    cacheShapeDiameter();
```
Comment (AndrewMeadows) [inline interface/src/avatar/AvatarMotionState.cpp]: This stuff about the "diameter" of the avatar is to fix a minor bug I noticed while investigating the motion of other avatars in the physics simulation.  The `_diameter` is used as a threshhold in`AvatarMotionState::setWorldTransform()` where we can slave the body's motion in two ways: slam its transform directly or set its velocity toward its target transform.

File: interface/src/avatar/AvatarMotionState.cpp [hunk 2]
```diff
@@ -141,7 +139,10 @@ glm::vec3 AvatarMotionState::getObjectLinearVelocity() const {
 
 // virtual
 glm::vec3 AvatarMotionState::getObjectAngularVelocity() const {
-    return _avatar->getWorldAngularVelocity();
+    // HACK: avatars use a capusle collision shape and their angularVelocity in the local simulation is unimportant.
+    // Therefore, as optimization toward support for larger crowds we ignore it and return zero.
+    //return _avatar->getWorldAngularVelocity();
+    return glm::vec3(0.0f);
```
Comment (AndrewMeadows) [inline interface/src/avatar/AvatarMotionState.cpp]: This is the main fix for the high avatar count test case where all bots were spinning in place.  By ignoring angular velocity the avatar's RigidBody is more likely to be deactivated in the physics simulation.  Note: one assumption here is that the other avatar's angular velocity is parallel to its capsule's axis of symmetry.

File: libraries/shared/src/PhysicsCollisionGroups.h [hunk 3]
```diff
@@ -59,7 +59,11 @@ const int32_t BULLET_COLLISION_MASK_KINEMATIC = BULLET_COLLISION_MASK_STATIC;
 // MY_AVATAR does not collide with itself
 const int32_t BULLET_COLLISION_MASK_MY_AVATAR = ~(BULLET_COLLISION_GROUP_COLLISIONLESS | BULLET_COLLISION_GROUP_MY_AVATAR);
 
-const int32_t BULLET_COLLISION_MASK_OTHER_AVATAR = BULLET_COLLISION_MASK_DEFAULT;
+// OTHER_AVATARs are dynamic, but are slammed to whatever the avatar-mixer says, which means
+// their motion can't actually be affected by the local physics simulation -- we rely on the remote simulation
+// to move its avatar around correctly and to communicate its motion through the avatar-mixer.
+// Therefore, they only need to collide against things that can be affected by their motion: dynamic and MyAvatar
+const int32_t BULLET_COLLISION_MASK_OTHER_AVATAR = BULLET_COLLISION_GROUP_DYNAMIC | BULLET_COLLISION_GROUP_MY_AVATAR;
```
Comment (AndrewMeadows) [inline libraries/shared/src/PhysicsCollisionGroups.h]: The other main fix is to adjust the collision mask of other avatar's to NOT include static or kinematic objects.   Due to the nature of the distributed physics simulation: collisions between other avatars and such objects doesn't actually matter.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11777/HighFidelity-Beta-PR13732-c939699.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11781/HighFidelity-Beta-PR13732-bd6f00e.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11781/HighFidelity-Beta-PR13732-bd6f00e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11781/HighFidelity-Beta-Interface-PR13732-bd6f00e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11781/HighFidelity-Beta-PR13732-bd6f00e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11781/HighFidelity-Beta-Interface-PR13732-bd6f00e.exe

**Comment by AndrewMeadows** [comment]
**Test Plan**:

**(1)** Find a domain with 200+ avatars in it

**(2)** Enable the following menu options:
**Developer --> Show Statistics**
**Developer --> Timing --> Performance Timer --> Display Timing Details**
**Developer --> Timing --> Performance Timer --> Expand /update**
**Developer --> Timing --> Performance Timer --> Expand /simulation**

**(3)** Look at for the **/idle/update/simulation/stepPhysics** stat
--> on `hifi/master` the **stepPhysics** stat may be at 15msec or higher and the **Game Rate** will be low (20?)
--> on this PR the **stepPhysics** stat will be lower 1.0 msec or less, depending on the content in the domain

**(4)** Make sure MyAvatar-vs-otherAvatar collisions still work.  Note: if you're pushing MyAvatar against a bot avatar they will not move at all, however if you push MyAvatar against another avatar driven by a real interface (which has its own physics simulation) you should be able to budge the otherAvatar a little bit.

**(5)** Make sure if you throw a dynamic object at an otherAvatar the object still bounces off the otherAvatar's collision shape in the physics engine.

**(6)** Enable the option **Developer --> Physics --> Show Bullet Collision** and verify each otherAvatar has a capsule collision shape that is located at about the right spot.  In a less busy domain have another interface login and walk its avatar around while you watch its physics collision capsule --> the collision shape should follow the moving avatar around.

File: interface/src/avatar/AvatarMotionState.cpp [hunk 4]
```diff
@@ -174,3 +179,31 @@ float AvatarMotionState::getMass() const {
...
+    ObjectMotionState::setRigidBody(body);
+    if (_body) {
+        // remove angular dynamics from this body
+        _body->setAngularFactor(0.0f);
+    }
+}
+
+void AvatarMotionState::setShape(const btCollisionShape* shape) {
+    cacheShapeDiameter();
+    ObjectMotionState::setShape(shape);
```
Comment (sabrina-shanman) [inline interface/src/avatar/AvatarMotionState.cpp]: cacheShapeDiameter should be called after ObjectMotionState::setShape

File: interface/src/avatar/AvatarMotionState.cpp (hunk 2 above)
Comment (SamGondelman) [reply interface/src/avatar/AvatarMotionState.cpp]: In the future, we could detect that this is the case.  I can imagine Avatars spinning with their pitch or roll that we would need to simulate.

File: interface/src/avatar/AvatarMotionState.cpp [hunk 5]
```diff
@@ -174,3 +179,31 @@ float AvatarMotionState::getMass() const {
...
+    if (_shape) {
+        // measure XZ diameter of capsule shape
+        btVector3 aabbMin, aabbMax;
+        btTransform transform;
+        transform.setIdentity();
+        _shape->getAabb(transform, aabbMin, aabbMax);
+        aabbMax -= aabbMin;
+        aabbMax.setY(0.0f);
+        const float SQRT_TWO = 1.414213562f;
+        _diameter = SQRT_TWO * aabbMax.length();
```
[the remark this answers is not in the document]
Comment (AndrewMeadows) [reply interface/src/avatar/AvatarMotionState.cpp]: Yeah I guess you're right.  This code is agnostic about the actual shape and provides an average radius in the case where the shape doesn't have radial symmetry.  However, we are already making the assumption elsewhere that the shape is a capsuleY  so we might as well also make it here.

File: interface/src/avatar/AvatarMotionState.cpp (hunk 4 above)
Comment (AndrewMeadows) [reply interface/src/avatar/AvatarMotionState.cpp]: D'oh!  Good eye.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11802/HighFidelity-Beta-PR13732-e336c92.apk

File: interface/src/avatar/AvatarMotionState.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (sabrina-shanman) [reply interface/src/avatar/AvatarMotionState.cpp]: After a second look, I think the aabb length is already effectively the hypotenuse of the x and z sides, although if it's working well now then the exact value doesn't matter too much.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11802/HighFidelity-Beta-PR13732-e336c92.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11802/HighFidelity-Beta-Interface-PR13732-e336c92.dmg

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by qlimnavarro** [comment]
Pass @conklin94122 
Numbers look good and line up with other versions' results

**Pull request merged**